wrappers/wrapper.rb

```diff
@@ -742,89 +748,69 @@ def detect_unfeasible_services(vrp)
+      matrix = vrp.matrices.find{ |mat| mat.id == vehicle.matrix_id }
 
-            if vehicle.cost_time_multiplier&.positive? &&
-               !vehicle.cost_late_multiplier&.positive? &&
-               ((vehicle.timewindow&.start && vehicle.timewindow&.end) || vehicle.sequence_timewindows&.size&.positive?)
-              vehicle_available_time = vehicle.timewindow.end - vehicle.timewindow.start if vehicle.timewindow
-              vehicle_available_time = vehicle.sequence_timewindows.collect{ |tw| tw.end - tw.start }.max if vehicle.sequence_timewindows&.size&.positive?
+      vehicle_start = vehicle.timewindow&.start || vehicle.sequence_timewindows.collect(&:start).min || 0
+      vehicle_end = vehicle.timewindow&.end || vehicle.sequence_timewindows.collect(&:end).max || 2**56
+      vehicle_available_time = vehicle_end - vehicle_start
 
-              min_time = (start_index && matrix[:time][start_index][index]).to_f + (end_index && matrix[:time][index][end_index]).to_f + activity.duration
-
-              feasible &&= (vehicle_available_time >= min_time)
-            end
+      earliest_arrival = vehicle_start
+      earliest_arrival += matrix.time[v_start_index][service_index] if v_start_index
+      if !activity.late_multiplier&.positive? && activity.timewindows.any?(&:end)
+        return false if activity.timewindows.all?{ |tw| tw.end && earliest_arrival > tw.end }
+      end
 
-            if feasible &&
-               (start_index || end_index) &&
-               vehicle.cost_distance_multiplier&.positive? &&
-               vehicle.distance
-              min_dist = (start_index && matrix[:distance][start_index][index]).to_f + (end_index && matrix[:distance][index][end_index]).to_f
+      if vehicle_available_time && vehicle.cost_time_multiplier&.positive? && !vehicle.cost_late_multiplier&.positive?
+        earliest_v_end = earliest_arrival + activity.duration
+        earliest_v_end += matrix.time[service_index][v_end_index] if v_end_index
+        return false if (earliest_v_end - vehicle_start) > vehicle_available_time
+      end
 
-              feasible &&= (vehicle.distance >= min_dist)
-            end
+      if !vehicle.cost_late_multiplier&.positive? && activity.timewindows.any?{ |tw| tw.start.positive? } &&
+         (vehicle.timewindow&.end || vehicle.sequence_timewindows.any?(&:end))
+        latest_arrival = vehicle_end - activity.duration
+        latest_arrival -= matrix[:time][service_index][vehicle.end_point.matrix_index] if vehicle.end_point_id
+        return false if activity.timewindows.all?{ |tw| latest_arrival < tw.start }
+      end
 
-            feasible
-          }
-        }
+      if v_start_index && v_end_index && vehicle.distance && vehicle.cost_distance_multiplier&.positive?
+        min_dist = matrix[:distance][v_start_index][service_index].to_f +
+                   matrix[:distance][service_index][v_end_index].to_f
+        return false if min_dist > vehicle.distance
+      end
```

Comment:
We can just do as follows because if both `v_start_index` and `v_end_index` are `nil` then the `min_distance` will be `0` and `vehicle.distance` cannot be negative but if one of them us non-nil then we want to check the distance!
```suggestion
      if vehicle.distance && vehicle.cost_distance_multiplier&.positive?
        min_dist = (v_start_index && matrix[:distance][v_start_index][service_index]).to_f +
                   (v_end_index && matrix[:distance][service_index][v_end_index]).to_f
        return false if min_dist > vehicle.distance
      end
```
which translates into to correct if condition `(v_start_index || v_end_index)` and not `v_start_index && v_end_index`